### PR TITLE
refactor(account-settings): share `<ValidationErrors />`

### PR DESCRIPTION
**Pull request by wlee221**
#### Description of changes

`Authenticator` and `ChangePassword` share very similar logic, especially regarding handling password validation. This PR shares password validation utils so that Account Settings do not have to duplicate these logics.

#### Details
- Share `defaultPasswordValidator` util, which checks password against password settings defined in zero config.
- Share `<ValidationErrors />` React component, which will render array of validation errors. 

#### Description of how you validated changes

Updated unit tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react/src/components/Authenticator/shared/FormField.tsx [hunk 1]
```diff
@@ -45,7 +45,11 @@ export function FormField({
           hasError={hasError}
           aria-describedby={ariaDescribedBy}
         />
-        <ValidationErrors errors={errors} id={errorId} />
+        <ValidationErrors
+          dataAttr="data-amplify-sign-up-errors"
```
Comment (wlee221) [inline packages/react/src/components/Authenticator/shared/FormField.tsx:49]: These `dataAttr` is specific to Authenticator, so I made this a prop.

File: packages/ui/src/helpers/shared/__tests__/validator.test.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
+import { PasswordSettings } from '../../../types';
+import { ALLOWED_SPECIAL_CHARACTERS } from '../../../helpers/authenticator/constants';
+import { defaultPasswordValidator } from '../validator';
+
+const strictPasswordPolicy: PasswordSettings = {
```
Comment (wlee221) [inline packages/ui/src/helpers/shared/__tests__/validator.test.ts:5]: These tests are migrated from `defaultServices.test.ts`.

File: packages/ui/src/helpers/shared/validator.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+export const defaultPasswordValidator = (
+  password: string,
+  passwordSettings: PasswordSettings
+): string[] => {
+  const errors = [];
+
+  const policyMinLength = passwordSettings?.passwordPolicyMinLength;
+  if (password.length < policyMinLength) {
+    errors.push(`Password must have at least ${policyMinLength} characters`);
+  }
```
Comment (wlee221) [inline packages/ui/src/helpers/shared/validator.ts:14]: These logics are migrated from `defaultServices.validateFormPassword`, minus Authenticator specific classnames or blur logic.

File: packages/react/src/components/Authenticator/shared/FormField.tsx (hunk 1 above)
Comment (ErikCH) [reply packages/react/src/components/Authenticator/shared/FormField.tsx:49]: Isn't `ValidationErrors` only used in the Authenticator? So shouldn't it always use the `data-amplify-sign-up-errors` prop? Or are there some scenarios where you wouldn't pass this in?

File: packages/ui/src/helpers/shared/validator.ts (hunk 3 above)
Comment (ErikCH) [reply packages/ui/src/helpers/shared/validator.ts:14]: Was all this moved just to clean up the `defaultServices` logic?

Comment (ErikCH) [reply packages/ui/src/helpers/shared/validator.ts:14]: I see and help with duplication of logic when the Account Services checks go into place. 👍

File: packages/react/src/components/Authenticator/shared/FormField.tsx (hunk 1 above)
Comment (ErikCH) [reply packages/react/src/components/Authenticator/shared/FormField.tsx:49]: Will the Account settings be using this Validation Errors without the prop?

**Review by ErikCH — APPROVED** [review]
Just a few questions... but looks fine

File: packages/ui/src/helpers/shared/validator.ts (hunk 3 above)
Comment (wlee221) [reply packages/ui/src/helpers/shared/validator.ts:14]: Right, I should've added more to PR description 😓 

This is specifically for `ChangePassword` component, which will have almost the same validation logic as Authenticator one.

File: packages/react/src/components/Authenticator/shared/FormField.tsx (hunk 1 above)
Comment (wlee221) [reply packages/react/src/components/Authenticator/shared/FormField.tsx:49]: They won't be for now, I don't see its styles to be too useful. Will discuss with @dbanksdesign when I make the PR (#2900) ready for review.

File: packages/react/src/components/shared/__tests__/ValidationErrors.test.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      <ValidationErrors errors={['mock error 1', 'mock error 2']} />
+    );
+    expect(container).toMatchSnapshot();
+    expect(await screen.findByText('mock error 1')).toBeDefined();
+    expect(await screen.findByText('mock error 2')).toBeDefined();
+  });
+
+  it('renders nothing if there are no errors', async () => {
+    const { container } = render(<ValidationErrors errors={[]} />);
+    expect(container).toMatchInlineSnapshot(`<div />`);
```
Comment (joebuono) [inline packages/react/src/components/shared/__tests__/ValidationErrors.test.tsx:17]: Shouldn't it return `null` if there are no errors?

File: packages/ui/src/helpers/shared/__tests__/validator.test.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  passwordPolicyMinLength: 8,
+};
+
+const lenientPasswordPolicy: PasswordSettings = {
+  passwordPolicyCharacters: [],
+  passwordPolicyMinLength: 4,
+};
+
+describe('validateFormPassword', () => {
+  it.only('validates as expected with valid password and strict password policy', async () => {
```
Comment (joebuono) [inline packages/ui/src/helpers/shared/__tests__/validator.test.ts:21]: Remove `.only`?

Comment (wlee221) [reply packages/ui/src/helpers/shared/__tests__/validator.test.ts:21]: Oh gosh, thanks for the catch.

File: packages/react/src/components/shared/__tests__/ValidationErrors.test.tsx (hunk 4 above)
Comment (wlee221) [reply packages/react/src/components/shared/__tests__/ValidationErrors.test.tsx:17]: Yeah, I was going back and forth about this. I was initially thinking about having `Authenticator` or `AccountSettings` handle empty array and make it `null`, but I think it makes sense to just return `null` here.

Adjusting.

**Review by joebuono — APPROVED** [review]

File: packages/ui/src/helpers/shared/__tests__/validator.test.ts (hunk 5 above)
Comment (wlee221) [reply packages/ui/src/helpers/shared/__tests__/validator.test.ts]: Updated in [6078356](https://github.com/aws-amplify/amplify-ui/pull/2908/commits/6078356e6e1184668aa1eee66e12b43b8abaa2b8) 🙏

File: packages/react/src/components/shared/__tests__/ValidationErrors.test.tsx (hunk 4 above)
Comment (wlee221) [reply packages/react/src/components/shared/__tests__/ValidationErrors.test.tsx:17]: Updated in [4e570a5](https://github.com/aws-amplify/amplify-ui/pull/2908/commits/4e570a53152c6c7469fda6088caff3370b75491f). Added one unhappy unit test to `defaultService.test.ts` to feel safe too.

**Review by joebuono — APPROVED** [review]

**Comment by wlee221** [comment]
I ended up removing `defaultPasswordValidator`, because it still had much divergence from AccountSettings.ChangePassword validation. Sorry for the hussle. 

`ValidationErrors` component is left untouched. Will merge tomorrow morning unless there are objections.

**Pull request merged**